### ISSUE TITLE
DeepSpeed Zero -3 with DPO Trainer for Llama3. 

### ISSUE THREAD

**Issue by anki08** [huggingface/trl#2492]
### System Info

**ENV**
- Platform: Linux-5.10.228-219.884.amzn2.x86_64-x86_64-with-glibc2.31
- Python version: 3.10.8
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA A10G, NVIDIA A10G, NVIDIA A10G, NVIDIA A10G, NVIDIA A10G, NVIDIA A10G, NVIDIA A10G, NVIDIA A10G
- Transformers version: 4.47.0
- Accelerate version: 1.2.1
- Accelerate config: 
  - compute_environment: LOCAL_MACHINE
  - distributed_type: DEEPSPEED
  - mixed_precision: fp16
  - use_cpu: False
  - debug: False
  - num_processes: 8
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - deepspeed_config: {'gradient_accumulation_steps': 4, 'offload_optimizer_device': 'none', 'offload_param_device': 'cpu', 'zero3_init_flag': True, 'zero3_save_16bit_model': False, 'zero_stage': 3}
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 3.2.0
- HF Hub version: 0.27.0
- TRL version: 0.13.0
- bitsandbytes version: 0.45.0
- DeepSpeed version: 0.15.4
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: not installed

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

```python
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import os
 

dataset = load_dataset("csv", data_files="...", split="train")

model_id = "meta-llama/Llama-3.1-8B-Instruct"
model = AutoModelForCausalLM.from_pretrained(model_id, low_cpu_mem_usage=True)
model_ref = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.1-8B-Instruct", low_cpu_mem_usage=True)
for param in model_ref.parameters():
    param.requires_grad = False

tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.padding = "[PAD]"
tokenizer.pad_token = tokenizer.eos_token

training_args = DPOConfig(
    output_dir="LLAMA_1", 
    max_length=500, 
    per_device_train_batch_size=1,
    per_device_eval_batch_size=4, 
    num_train_epochs=1, 
    gradient_accumulation_steps=4, 
    fp16=True,
    optim="sgd",
    logging_steps=1,
    save_strategy= "epoch",
    save_steps= 5,
    run_name="SQL_Thinking_llms",
    learning_rate= 1e-5,
    gradient_checkpointing=True)
trainer = DPOTrainer(model=model, ref_model=model_ref, args=training_args, tokenizer=tokenizer, train_dataset=dataset)
trainer.train().

```

outputs:

```
{'loss': 0.6934, 'grad_norm': 0.0, 'learning_rate': 6.666666666666667e-06, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': 0.0, 'logps/rejected': 0.0, 'logits/chosen': nan, 'logits/rejected': nan, 'epoch': 0.33}                     
 33%|███████████████████████████████████████████████████████████████████████████████████████▋                                                                                                                                                                               | 1/3 [03:43<07:27, 223.67s/it][2024-12-16 23:46:06,264] [WARNING] [stage3.py:2114:step] 9 pytorch allocator cache flushes since last step. this happens when there is high memory pressure and is detrimental to performance. if this is happening frequently consider adjusting settings to reduce memory consumption. If you are unable to make the cache flushes go away consider adding get_accelerator().empty_cache() calls in your training loop to ensure that all ranks flush their caches at the same time
{'loss': 0.6934, 'grad_norm': 0.0, 'learning_rate': 3.3333333333333333e-06, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': 0.0, 'logps/rejected': 0.0, 'logits/chosen': nan, 'logits/rejected': nan, 'epoch': 0.67}                    
 67%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                                                                       | 2/3 [07:14<03:36, 216.08s/it][2024-12-16 23:49:36,778] [WARNING] [stage3.py:2114:step] 9 pytorch allocator cache flushes since last step. this happens when there is high memory pressure and is detrimental to performance. if this is happening frequently consider adjusting settings to reduce memory consumption. If you are unable to make the cache flushes go away consider adding get_accelerator().empty_cache() calls in your training loop to ensure that all ranks flush their caches at the same time
{'loss': 0.6934, 'grad_norm': 0.0, 'learning_rate': 0.0, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': 0.0, 'logps/rejected': 0.0, 'logits/chosen': nan, 'logits/rejected': nan, 'epoch': 1.0}                                        
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [10:44<00:00, 213.54s/it]
    ...
```

### Expected behavior

I am trying to use DPOTrainer to finetune llama-3-instruct. I am using deepspeed-zero-3. My dataset is a csv file consisting of 3 columns -> prompts, chosen and rejected. 
However,
The loss is not decreasing and the logits are nan. logits/chosen': nan, 'logits/rejected': nan and 
the logp are 0. the rewards are also 0.

### Checklist

- [X] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [X] I have included my system information
- [X] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any traceback provided is complete
